### PR TITLE
AIWorldPathfinder: add support for paving the way through Whirlpools

### PR DESCRIPTION
**Pull request by oleg-derevenetz**
related to #4509

Continuation of #4894. That's how it currently works ([WRLTEST.ZIP](https://github.com/ihhub/fheroes2/files/7824054/WRLTEST.ZIP)):

Path through the whirlpool

The `AIWorldPathfinder::reEvaluateIfNeeded()` now has optional parameter `considerWhirlpools` (`false` by default).

The debugging principle is just like in #4894 (right-click by the mouse when debug hero is selected).

Please note that `World::GetWhirlpoolEndPoints()` has been rewritten and now it is perfectly deterministic, so it can be used just like `World::GetTeleportEndPoints()` in the AI-related code.

BTW one more funny note: since whirlpools are multi-tile objects, the pathfinder currently chooses the "best tile" of the whirlpool for diving and surfacing from the water.  This leads to the following trick: instead of going around the whirlpool, pathfinder may (and will) choose to dive from the one side of the whirlpool and then emerge from the other side, thus spending 0 move points while traversing the whirlpool :)

Hi @ihhub what do you think?

### CODE REVIEW

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , sadly we have merge conflicts here 😞 I'll take a look at this pull request later today.

**Comment by oleg-derevenetz** [comment]
@ihhub No problem, I'll rebase this now.

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , it makes sense for AI to exploit such trick if it benefits from it.

**Comment by oleg-derevenetz** [comment]
> it makes sense for AI to exploit such trick if it benefits from it.

Yeah, but unfortunately human player can't do that, because in his case the surfacing tile is chosen randomly :)

**Comment by oleg-derevenetz** [comment]
@ihhub which would be better - merge this PR just with new `AIPathfinder`functionality (and implement whirlpool-related improvements for AI in a separate PR) or do everything within one PR?

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , since this pull request is specific only to a certain object on the map we should make all changes in one shot. I'll check how we can make AI to be smart enough to use both ways.

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , I've added simplest logic to determine if AI hero is able to pass throw whirlpools.

**Comment by oleg-derevenetz** [comment]
Hi @ihhub I added the logic to handle Whirlpools in the `AIToWhirlpool()` just like Teleports (to preserve the path while going through the Whirlpool), and if AI sees the whole map, it attacks my castle immediately on the test map (as soon as AI is able to reach it). However, if AI doesn't see the whole map, it shows no interest in the whirlpool - it just visits a nearby island by boat and that's all. Maybe we need to tweak something else?

**Comment by oleg-derevenetz** [comment]
BTW with the savefile from [here](https://github.com/ihhub/fheroes2/pull/4894#issuecomment-1006520512) AI still behaves the same way - endlessly dives into the whirlpool.

**Comment by ihhub** [comment]
> Hi @ihhub I believe that I have managed to solve all the problems related to the whirlpool here. Key ideas are:
> 
>     * Enable the fog discovery through the stone liths and whirlpools in a transparent manner;
> 
>     * Do not consider stone liths and whirlpools as objects from an AI perspective (just like boats);
> 
>     * Pre-arrange AI army to minimize whirlpool losses;
> 
>     * Considering the previous point, I propose to allow all hero types to use whirlpools, but it's optional and debatable, and may be reverted.
> 
> 
> Could you please take a look? As I can see, AI behaves well on all test maps, you can try it yourself. The only minor issue is when AI tries to pave the way through the whirlpool and gets to a wrong tile (because whirlpools are unpredictable) - in such cases it tries again, and this can happen multiple times.

Finally I have some free time. I'm looking right now.

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , I've tested multiple scenarios. AI heroes are still obsessed with whirlpools but only to pave the path. To reduce this obsession we can introduce a penalty from moving from one cell of a whirlpool to another one. But this can be done in a separate pull request.

**Comment by ihhub** [comment]
@oleg-derevenetz , also regarding allowance of every AI hero to use whirlpools: AI heroes has classes and not every hero should be allowed to use whirlpool and a very strong hero might loose a chunk of an important army. However, if it's easy to change in the future we can leave it as it is for now.

File: src/fheroes2/world/world_pathfinding.cpp [hunk 1]
```diff
@@ -531,60 +533,94 @@ int AIWorldPathfinder::getFogDiscoveryTile( const Heroes & hero )
...
+            if ( tilesVisited[newIndex] ) {
+                continue;
+            }
+
+            tilesVisited[newIndex] = true;
+
+            const MP2::MapObjectType newTileObject = world.GetTiles( newIndex ).GetObject( true );
+
+            // Don't go onto action objects (except Stone Liths and Whirlpools) as they might be castles or dwellings with guards.
+            if ( MP2::isActionObject( newTileObject ) && newTileObject != MP2::OBJ_STONELITHS && newTileObject != MP2::OBJ_WHIRLPOOL ) {
```
Comment (ihhub) [inline src/fheroes2/world/world_pathfinding.cpp:551]: Should we then add a Sign object as well? At least we can add a TODO comment to add "allowed" objects here.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @oleg-derevenetz , everything looks fine except few small issues which I pointed out. Could you please take a look at them?

File: src/fheroes2/world/world_pathfinding.cpp [hunk 2]
```diff
@@ -264,30 +265,24 @@ void PlayerWorldPathfinder::reset()
...
-        _pathStart = startIndex;
-        _currentColor = color;
-        _pathfindingSkill = skill;
-        _remainingMovePoints = remainingMovePoints;
-        _maxMovePoints = maxMovePoints;
-
-        processWorldMap( startIndex );
+    auto currentSettings = std::forward_as_tuple( _pathStart, _pathfindingSkill, _currentColor, _remainingMovePoints, _maxMovePoints );
+    const auto newSettings = std::make_tuple( hero.GetIndex(), static_cast<uint8_t>( hero.GetLevelSkill( Skill::Secondary::PATHFINDING ) ), hero.GetColor(),
+                                              hero.GetMovePoints(), hero.GetMaxMovePoints() );
```
[the remark this answers is not in the document]
Comment (oleg-derevenetz) [reply src/fheroes2/world/world_pathfinding.cpp:280]: @ihhub We did the same copying before, just like this:

```cpp
const int startIndex = hero.GetIndex();
const int color = hero.GetColor();
const double armyStrength = hero.GetArmy().GetStrength();
const uint32_t skill = hero.GetLevelSkill( Skill::Secondary::PATHFINDING );
const uint32_t remainingMovePoints = hero.GetMovePoints();
const uint32_t maxMovePoints = hero.GetMaxMovePoints();
```

The difference is that now we mention each parameter being compared exactly once, and there is no way to compare it but forget to update and so on.

File: src/fheroes2/ai/ai_hero_action.cpp [hunk 3]
```diff
@@ -927,63 +908,57 @@ namespace AI
...
-                    }
-                    return;
-                }
-                else if ( !other_hero->isFreeman() ) {
-                    DEBUG_LOG( DBG_GAME, DBG_WARN, other_hero->String() + " hero is blocking teleporter exit" );
-                    return;
-                }
-            }
-        }
+        // TODO: remove this assertion, this should never happen
```
[the remark this answers is not in the document]
Comment (oleg-derevenetz) [reply src/fheroes2/ai/ai_hero_action.cpp:911]: Discussed above.

File: src/fheroes2/world/world_pathfinding.cpp [hunk 4]
```diff
@@ -360,54 +357,45 @@ void AIWorldPathfinder::reset()
...
-         || _remainingMovePoints != remainingMovePoints || _maxMovePoints != maxMovePoints ) {
-        _pathStart = startIndex;
-        _currentColor = color;
-        _armyStrength = armyStrength;
-        _pathfindingSkill = skill;
-        _remainingMovePoints = remainingMovePoints;
-        _maxMovePoints = maxMovePoints;
-
-        processWorldMap( startIndex );
+    auto currentSettings = std::forward_as_tuple( _pathStart, _pathfindingSkill, _currentColor, _remainingMovePoints, _maxMovePoints, _armyStrength );
```
[the remark this answers is not in the document]
Comment (oleg-derevenetz) [reply src/fheroes2/world/world_pathfinding.cpp:372]: Discussed above.

**Comment by oleg-derevenetz** [comment]
Hi @ihhub

>  I've tested multiple scenarios. AI heroes are still obsessed with whirlpools but only to pave the path. To reduce this obsession we can introduce a penalty from moving from one cell of a whirlpool to another one. But this can be done in a separate pull request.

Most of the obsession is related to scenarios like this: consider that AI-controlled hero needs to get to the point marked by the cross:



and AI thinks that this can be done like this (and yes, it can be done like this - with a certain amount of luck):



But then happens this:



and AI is like this: "OMG, what to do? I just did the task recalculation, and I still need to get to that point marked by the cross! I will try once again." And then he comes up again in the wrong whirlpool. Issues like this can't be reliably mitigated by penalties. We can try to mitigate it by cheating (no random for AI in whirlpools and stone liths - because stone liths with multiple exits are vulnerable to this as well), but I prefer AI to suffer the same way as a human in this regard. Anyway, this can be discussed separately.

>  also regarding allowance of every AI hero to use whirlpools: AI heroes has classes and not every hero should be allowed to use whirlpool and a very strong hero might loose a chunk of an important army. However, if it's easy to change in the future we can leave it as it is for now.

The problem here is that a strong AI-controlled hero might hesitate to capture the unguarded castle (or eliminate the weak enemy hero) because of his fear of whirlpools. I saw this even on the test map - the hero with the Titans did not want to attack the unguarded castle because of the whirlpool. That's why I added that logic instead, that does the auto-arrangement of the hero's army before the whirlpool (or, technically, before calculating losses) to minimize that losses.

File: src/fheroes2/heroes/heroes_action.cpp [hunk 5]
```diff
@@ -1991,19 +1991,8 @@ void ActionToTeleports( Heroes & hero, s32 index_from )
...
-        // lose battle
-        if ( hero.isFreeman() ) {
-            return;
-        }
-        else if ( !other_hero->isFreeman() ) {
-            DEBUG_LOG( DBG_GAME, DBG_WARN, "is busy..." );
-            return;
-        }
-    }
+    // TODO: remove this assertion, this should never happen
```
[the remark this answers is not in the document]
Comment (ihhub) [reply src/fheroes2/heroes/heroes_action.cpp:1994]: I'm okay to leave the assertion as it is to make sure that someone else in the future won't break the logic. Let's leave it as it is🙂

File: src/fheroes2/world/world_pathfinding.cpp (hunk 1 above)
Comment (ihhub) [reply src/fheroes2/world/world_pathfinding.cpp:551]: There could be a scenario where the only way to discover more area is to pass through a sign.

Comment (ihhub) [reply src/fheroes2/world/world_pathfinding.cpp:551]: Let's leave it as it is for now.

**Review by ihhub — APPROVED** [review]

**Comment by ihhub** [comment]
@oleg-derevenetz , looks good to me. If you have no more changes pending then we can proceed with merging.

**Comment by oleg-derevenetz** [comment]
Hi @ihhub 

> looks good to me. If you have no more changes pending then we can proceed with merging.

No changes pending so far, if there are new ideas, it will be possible to make separate PRs :)

**Comment by ihhub** [comment]
@oleg-derevenetz , wonderful job! Well done!

**Pull request merged**